File: backtest/okx.mjs
const OKX_HISTORY_CANDLES_URL = "https://www.okx.com/api/v5/market/history-candles";
const DEFAULT_REQUEST_LIMIT = 100;
const DAY_MS = 24 * 60 * 60 * 1000;
const HOUR_MS = 60 * 60 * 1000;
const PAGE_SAFETY_MULTIPLIER = 1.2;
const PAGE_SAFETY_EXTRA = 3;
const RETRY_DELAYS_MS = [2000, 4000, 8000, 16000];

function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function rowTimestamp(row) {
  return Number(row?.[0]);
}

function barToMs(bar) {
  const text = String(bar || "1D");
  const value = Number.parseInt(text, 10) || 1;

  if (text.endsWith("m")) return value * 60 * 1000;
  if (text.endsWith("H")) return value * HOUR_MS;
  if (text.endsWith("D")) return value * DAY_MS;
  if (text.endsWith("W")) return value * 7 * DAY_MS;
  return DAY_MS;
}

function calculateMaxPages(config) {
  const barMs = barToMs(config.bar);
  const requestLimit = Math.max(1, Math.floor(Number(config.requestLimit) || DEFAULT_REQUEST_LIMIT));
  const requestedBars = Math.ceil((config.days * DAY_MS) / barMs);

  return {
    requestLimit,
    maxPages: Math.max(1, Math.ceil((requestedBars / requestLimit) * PAGE_SAFETY_MULTIPLIER) + PAGE_SAFETY_EXTRA)
  };
}

function timestampToIso(timestamp) {
  return Number.isFinite(timestamp) ? new Date(timestamp).toISOString() : null;
}

async function fetchOkxPage(url) {
  let lastError = null;

  for (let attempt = 0; attempt <= RETRY_DELAYS_MS.length; attempt += 1) {
    try {
      const response = await fetch(url.toString(), { cache: "no-store" });
      if (!response.ok) {
        throw new Error(`OKX request failed: ${response.status} ${response.statusText}`);
      }

      const payload = await response.json();
      if (payload.code !== "0" || !Array.isArray(payload.data)) {
        throw new Error(`OKX response unavailable: ${payload.msg || payload.code || "unknown"}`);
      }

      return {
        payload,
        retryCount: attempt
      };
    } catch (error) {
      lastError = error;
      if (attempt >= RETRY_DELAYS_MS.length) break;
      await sleep(RETRY_DELAYS_MS[attempt]);
    }
  }

  throw lastError;
}

export async function downloadOkxHistory(config) {
  const startedAt = new Date();
  const startMs = startedAt.getTime() - (config.days * 24 * 60 * 60 * 1000);
  const { requestLimit, maxPages } = calculateMaxPages(config);
  const rowsByTime = new Map();
  let cursor = null;
  let page = 0;
  let lastOldest = Infinity;
  let oldestReached = null;
  let retryCount = 0;

  while (page < maxPages) {
    const url = new URL(OKX_HISTORY_CANDLES_URL);
    url.searchParams.set("instId", config.instrument);
    url.searchParams.set("bar", config.bar);
    url.searchParams.set("limit", String(requestLimit));
    if (cursor) url.searchParams.set("after", cursor);

    const pageResult = await fetchOkxPage(url);
    retryCount += pageResult.retryCount;

    const pageRows = pageResult.payload.data.filter((row) => Number.isFinite(rowTimestamp(row)));
    if (!pageRows.length) break;

    for (const row of pageRows) {
      rowsByTime.set(String(rowTimestamp(row)), row);
    }

    const oldest = Math.min(...pageRows.map(rowTimestamp));
    oldestReached = oldestReached === null ? oldest : Math.min(oldestReached, oldest);
    page += 1;

    if (oldest <= startMs || oldest >= lastOldest) break;
    lastOldest = oldest;
    cursor = String(oldest);
    await sleep(140);
  }

  const rows = Array.from(rowsByTime.values())
    .sort((left, right) => rowTimestamp(left) - rowTimestamp(right))
    .filter((row) => rowTimestamp(row) >= startMs);
  const truncated = oldestReached === null || oldestReached > startMs;

  return {
    source: "OKX",
    endpoint: OKX_HISTORY_CANDLES_URL,
    instrument: config.instrument,
    bar: config.bar,
    requestedDays: config.days,
    requestedStartMs: startMs,
    requestedStartDate: timestampToIso(startMs),
    requestLimit,
    maxPages,
    downloadedAt: new Date().toISOString(),
    pageCount: page,
    rowCount: rows.length,
    retryCount,
    oldestReached,
    oldestReachedDate: timestampToIso(oldestReached),
    truncated,
    rows
  };
}

File: backend_py/research/okx.py
from __future__ import annotations

from datetime import datetime, timezone
import json
import math
import time
from typing import Any
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from backend_py.research.clean import DAY_MS, HOUR_MS, bar_to_ms
from backend_py.research.config import ResearchConfig


OKX_HISTORY_CANDLES_URL = "https://www.okx.com/api/v5/market/history-candles"
DEFAULT_REQUEST_LIMIT = 100
PAGE_SAFETY_MULTIPLIER = 1.2
PAGE_SAFETY_EXTRA = 3
RETRY_DELAYS_SECONDS = [2, 4, 8, 16]


def timestamp_to_iso(timestamp: int | float | None) -> str | None:
    if timestamp is None or not math.isfinite(float(timestamp)):
        return None
    return datetime.fromtimestamp(float(timestamp) / 1000, tz=timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def row_timestamp(row: Any) -> float:
    try:
        return float(row[0])
    except (TypeError, ValueError, IndexError):
        return float("nan")


def calculate_max_pages(config: ResearchConfig) -> dict[str, int]:
    request_limit = max(1, math.floor(float(config.requestLimit or DEFAULT_REQUEST_LIMIT)))
    requested_bars = math.ceil((config.days * DAY_MS) / bar_to_ms(config.bar))
    max_pages = max(1, math.ceil((requested_bars / request_limit) * PAGE_SAFETY_MULTIPLIER) + PAGE_SAFETY_EXTRA)
    return {"requestLimit": request_limit, "maxPages": max_pages}


def fetch_okx_page(url: str) -> dict[str, Any]:
    last_error: Exception | None = None
    for attempt in range(len(RETRY_DELAYS_SECONDS) + 1):
        try:
            request = Request(url, headers={"User-Agent": "whd-python-research/1.0"})
            with urlopen(request, timeout=30) as response:
                payload = json.loads(response.read().decode("utf-8"))
            if payload.get("code") != "0" or not isinstance(payload.get("data"), list):
                raise RuntimeError(f"OKX response unavailable: {payload.get('msg') or payload.get('code') or 'unknown'}")
            return {"payload": payload, "retryCount": attempt}
        except Exception as error:  # noqa: BLE001 - retry wrapper returns final upstream error.
            last_error = error
            if attempt >= len(RETRY_DELAYS_SECONDS):
                break
            time.sleep(RETRY_DELAYS_SECONDS[attempt])
    raise RuntimeError(str(last_error))


def download_okx_history(config: ResearchConfig) -> dict[str, Any]:
    started_at = datetime.now(timezone.utc)
    start_ms = int(started_at.timestamp() * 1000) - config.days * HOUR_MS * 24
    page_config = calculate_max_pages(config)
    request_limit = page_config["requestLimit"]
    max_pages = page_config["maxPages"]
    rows_by_time: dict[str, list[Any]] = {}
    cursor: str | None = None
    page = 0
    last_oldest = float("inf")
    oldest_reached: float | None = None
    retry_count = 0

    while page < max_pages:
        query = {"instId": config.instrument, "bar": config.bar, "limit": str(request_limit)}
        if cursor:
            query["after"] = cursor
        page_result = fetch_okx_page(f"{OKX_HISTORY_CANDLES_URL}?{urlencode(query)}")
        retry_count += int(page_result["retryCount"])

        page_rows = [row for row in page_result["payload"]["data"] if math.isfinite(row_timestamp(row))]
        if not page_rows:
            break

        for row in page_rows:
            rows_by_time[str(int(row_timestamp(row)))] = row

        oldest = min(row_timestamp(row) for row in page_rows)
        oldest_reached = oldest if oldest_reached is None else min(oldest_reached, oldest)
        page += 1

        if oldest <= start_ms or oldest >= last_oldest:
            break
        last_oldest = oldest
        cursor = str(int(oldest))
        time.sleep(0.14)

    rows = [
        row
        for row in sorted(rows_by_time.values(), key=row_timestamp)
        if row_timestamp(row) >= start_ms
    ]
    truncated = oldest_reached is None or oldest_reached > start_ms
    return {
        "source": "OKX",
        "endpoint": OKX_HISTORY_CANDLES_URL,
        "instrument": config.instrument,
        "bar": config.bar,
        "requestedDays": config.days,
        "requestedStartMs": start_ms,
        "requestedStartDate": timestamp_to_iso(start_ms),
        "requestLimit": request_limit,
        "maxPages": max_pages,
        "downloadedAt": utc_now_iso(),
        "pageCount": page,
        "rowCount": len(rows),
        "retryCount": retry_count,
        "oldestReached": int(oldest_reached) if oldest_reached is not None else None,
        "oldestReachedDate": timestamp_to_iso(oldest_reached),
        "truncated": truncated,
        "rows": rows,
    }
